tutorials/arvo/parsing.md

```diff
@@ -0,0 +1,638 @@
+`q.q.edg` is `"bc"`, which is the part of the input `tape` that has yet to be parsed.
+
+Now let's see what happens when the parse fails.
+```
+> =edg ((just 'b') [[1 1] "abc"])
+> edg
+[p=[p=1 q=1] q=~]
+```
+Now we have that `p.edg` is the same as the input `hair`, `[1 1]`, meaning the
+parser has not advanced since the parse failed. `q.edg` is null, indicating that
+the parse has failed.
+
+Later we will use [+star](#star) to string together a sequence of `+just`s in
+order to parse multiple characters at once.
+
+### `+jest`
+
+`+jest` is a `rule` builder used to match a `cord`. It takes an input `cord` and
+produces a `rule` that attempts to match that `cord` against the beginning of
+the `tape` in an input `nail`.
+
+Let's see what happens when we successfully parse the entire input `tape`.
+```
+> =edg ((jest 'abc') [[1 1] "abc"])
+> edg
+[p=[p=1 q=4] q=[~ [p='abc' q=[p=[p=1 q=4] q=""]]]]
+```
+`p.edg` is `[p=1 q=4]`, indicating that the next character to be parsed is at
+line 1, column 4. Of course, this does not exist since the input `tape` was only
+3 characters long, so this actually indicates that the entire `tape` has been
+successfully parsed (since the `hair` does not advance in the case of failure).
+`p.q.edg` is `'abc'`, as expected. `q.q.edg` is `""`, indicating that nothing
+remains to be parsed.
+
+What happens if we only match some of the input `tape`?
+```
+> =edg ((jest 'ab') [[1 1] "abc"])
+> edg
+[p=[p=1 q=3] q=[~ [p='ab' q=[p=[p=1 q=3] q="c"]]]]
+```
+Now we have that the result, `p.q.edg`, is `'ab'`, while the remainder `q.q.q.edg`
+is `"c"`. So `+jest` has successfully parsed the first two characters, while the
+last character remains. Furthermore, we still have the information that the
+remaining character was in line 1 column 3 from `p.edg` and `p.q.q.edg`.
+
+What happens when `+jest` fails?
+```
+> ((jest 'bc') [[1 1] "abc"])
+[p=[p=1 q=1] q=~]
+```
+Despite the fact that `'bc'` appears in `"abc"`, because it was not at the
+beginning the parse failed. We will see in [parser
+combinators](#parser-combinators) how to modify this `rule` so that it
+finds `bc` successfully.
+
+### `+shim`
+
+`+shim` is used to parse characters within a given range. It takes in two atoms
+and returns a `rule`.
+
+```
+> ((shim 'a' 'z') [[1 1] "abc"])
+[p=[p=1 q=2] q=[~ [p='a' q=[p=[p=1 q=2] q="bc"]]]]
+```
+
+### `+cold`
+
+`+cold` is a `rule` builder that takes in a constant noun we'll call `cus` and a
+`rule` we'll call `sef` and returns a `rule` identical to the `sef` except it
+returns `cus` when it parses successfully.
+
+Here we see that `p.q` of the `edge` returned by the `rule` created with `+cold`
+is `%foo`.
+```
+> ((cold %foo (just 'a')) [[1 1] "abc"])
+[p=[p=1 q=2] q=[~ u=[p=%foo q=[p=[p=1 q=2] q="bc"]]]]
+```
+
+One common scenario where `+cold` sees play is when writing [command line
+interface (CLI) apps](@/docs/tutorials/arvo/cli-tutorial.md). We usher the
+reader there to find an example where `+cold` is used.
```

Comment:
This feels a little awkward, and I think in general you want links to flow the other way. Right now you have a primary resource pointing to a secondary one. This page is about parsers, the CLI tutorial isn't. If anything, as far as parsers are concerned, the CLI tutorial should be linking here!

Reply:
The CLI tutorial does link here (see the followup PR I made shortly after this one) and I think bidirectional links are a good thing for community knowledge bases where you don't know what direction someone will be coming from. Single-direction linking is better for personal knowledge bases where using bidirectional links becomes a larger and larger task over time as the number of resources increases. The time spent on that is only justified when people other than yourself will be using it, as is the case here. There's a lot of theory about this if you look into the literature on knowledge management.